src/libs/decoders/mp3.cpp

```diff
@@ -0,0 +1,222 @@
+#  include <config.h>
+#endif
+
+#include <assert.h>
+
+#include <SDL.h> // provides: SDL_malloc, SDL_realloc, SDL_free, SDL_memcpy, and SDL_memset
+#define DR_MP3_IMPLEMENTATION
+#define DR_MP3_NO_STDIO 1
+#define DRMP3_ASSERT(x) assert((x))
+#define DRMP3_MALLOC(sz) SDL_malloc((sz))
+#define DRMP3_REALLOC(p, sz) SDL_realloc((p), (sz))
+#define DRMP3_FREE(p) SDL_free((p))
+#define DRMP3_COPY_MEMORY(dst, src, sz) SDL_memcpy((dst), (src), (sz))
+#define DRMP3_ZERO_MEMORY(p, sz) SDL_memset((p), 0, (sz))
+#include "dr_mp3.h" // provides: drmp3
+
+#include "mp3_seek_table.h" // provides: populate_seek_table and SDL_Sound headers
+
+#include "SDL_sound.h"
+#define __SDL_SOUND_INTERNAL__
+#include "SDL_sound_internal.h" // provides: Sound_SampleInternal
+
+#define MP3_FAST_SEEK_FILENAME "fastseek.lut"
+
+static size_t mp3_read(void* const pUserData, void* const pBufferOut, const size_t bytesToRead)
+{
+    Uint8* ptr = static_cast<Uint8*>(pBufferOut);
+    Sound_Sample* const sample = static_cast<Sound_Sample* const>(pUserData);
+    const Sound_SampleInternal* const internal = static_cast<const Sound_SampleInternal*>(sample->opaque);
+    SDL_RWops* rwops = internal->rw;
+    size_t retval = 0;
+
+    while (retval < bytesToRead)
+    {
+        const size_t rc = SDL_RWread(rwops, ptr, 1, bytesToRead);
+        if (rc == 0)
+        {
+            sample->flags |= SOUND_SAMPLEFLAG_EOF;
+            break;
+        } /* if */
+        else
+        {
+            retval += rc;
+            ptr += rc;
+        } /* else */
+    } /* while */
+
+    return retval;
+} /* mp3_read */
+
+static drmp3_bool32 mp3_seek(void* const pUserData, const Sint32 offset, const drmp3_seek_origin origin)
+{
+    const Sint32 whence = (origin == drmp3_seek_origin_start) ? RW_SEEK_SET : RW_SEEK_CUR;
+    Sound_Sample* const sample = static_cast<Sound_Sample*>(pUserData);
+    Sound_SampleInternal* const internal = static_cast<Sound_SampleInternal*>(sample->opaque);
+    return (SDL_RWseek(internal->rw, offset, whence) != -1) ? DRMP3_TRUE : DRMP3_FALSE;
+} /* mp3_seek */
+
+
+static Sint32 MP3_init(void)
+{
+    return 1;  /* always succeeds. */
+} /* MP3_init */
+
+
+static void MP3_quit(void)
+{
+    /* it's a no-op. */
+} /* MP3_quit */
+
+static void MP3_close(Sound_Sample* const sample)
+{
+    Sound_SampleInternal* const internal = static_cast<Sound_SampleInternal* const>(sample->opaque);
+    mp3_t* p_mp3 = static_cast<mp3_t*>(internal->decoder_private);
+    if (p_mp3 != NULL) {
+        if (p_mp3->p_dr != NULL) {
+            drmp3_uninit(p_mp3->p_dr);
+            SDL_free(p_mp3->p_dr);
+        }
+        // maps and vector destructors free their memory
+        SDL_free(p_mp3);
+        internal->decoder_private = NULL;
+    }
+} /* MP3_close */
+
+static Uint32 MP3_read(Sound_Sample* const sample)
+{
+    Sound_SampleInternal* const internal = static_cast<Sound_SampleInternal* const>(sample->opaque);
+    const Sint32 channels = (int) sample->actual.channels;
```

Comment:
Maybe better use Sint32 in cast.

Reply:
Yes!

Reply:
Fixed in https://github.com/dreamer/dosbox-staging/commit/c4b303a2414482fa5a4a02d3c094e83fba32dcf4